Add new Subaru integration

G-Two opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->


## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Adds new integration for Subaru vehicles equipped with active Starlink remote services subscriptions.  This initial PR includes only sensor entities to maintain simplicity and demonstrate the ability to receive vehicle information.  This integration uses the `subarulink` package from PyPI.
- Includes only sensor platform for this initial PR
- binary_sensor, device_tracker and switch platforms are ready to add following this PR
- services for remote start and lights/horn/locks are ready to add following this PR

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests


## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->
A valid MySubaru account and active vehicle subscription to Starlink remote services is required to use this integration.
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/13493

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

G-Two committed:
Initial PR for new subaru integration

- Includes only sensor and lock entities for this initial PR
- binary_sensor, climate, device_tracker and switch entities are ready to add following this PR

G-Two committed:
Fixed config flow tests.

G-Two committed:
Fixes pylint and py38 test failures.

- Adds additional config flow and option flow tests.

G-Two committed:
Adds subaru files .coveragerc

MatthewFlamm commented on homeassistant/components/subaru/__init__.py:
```diff
@@ -0,0 +1,173 @@
+    )
+    if unload_ok:
+        hass.data[DOMAIN].pop(entry.entry_id)
+
```
Should the DOMAIN be removed if all entries are removed?

G-Two replied:
Reverted. DOMAIN is no longer removed since it is assumed to be in place by async_setup().  Otherwise if an entry is deleted and a new one is added, the DOMAIN won't be there.

MatthewFlamm commented on homeassistant/components/subaru/__init__.py:
```diff
@@ -0,0 +1,173 @@
+        """Update the state of the device."""
+        try:
+            await self.subaru_device.async_update()
+        except SubaruException as ex:
```
It looks like you only need to do one update for all entities, consider using this framework with DataUpdateCoordinator https://developers.home-assistant.io/docs/integration_fetching_data#coordinated-single-api-poll-for-data-for-all-entities

G-Two replied:
@MatthewFlamm thanks for the review and comment.  I've been trying to use the DataUpdateCoordinator this week, but I don't think it will work for the current implementation of the subarulink package. It was designed to hide some of the messiness of the Subaru API, especially when dealing with multiple vehicles on one account (it requires multiple API calls to switch vehicle context and then fetch the data update).  The package is caching, so all the extraneous calls for the different entities won't result in any unnecessary API calls being made to Subaru services.  I think the implementation of subarulink.hass needs to eventually go into homeassistant and I'll make that a design goal for a future revision in order to take advantage of the various homeassistant data handling frameworks.

G-Two replied:
Integration has been updated to use DataUpdateCoordinator.  Thanks for the suggestion, this will make the addition of more sensors more efficient.

G-Two committed:
Removes DOMAIN if all subaru entries are removed.

G-Two committed:
Uses hass.config.units for sensor conversions

Updated to subarulink==0.2.2

G-Two commented:
I'm making significant changes to the underlying PyPI package to enhance future maintainability.  Once changes are completed, I'll mark as ready for review.

G-Two committed:
Removes subarulink.hass dependency

Removes Home Assistant integration code from subarulink
Simplifies entity inheritance
Uses DataUpdateCoordinator
Updated to subarulink 0.3.0

G-Two committed:
Fixes bug in FuelEconomySensor

G-Two committed:
Fixes incorrect use of DataUpdateCoordinator

stboch commented:
@G-Two Look at the latest update and review some things, glad you refactored the subarulink makes it much easier to support. I noticed some of the other platforms need some work, and I was looking and thinking how the other connected car's do most of theirs, and would suggest making most of them services. 

The only thing that I was going back and forth on was Remote Start/Climate whether they just should be a service to start and stop with the long choice of climate parameters.

G-Two committed:
Updates integration to use subarulink 0.3.1

Geezus42 commented:
Is there anything I could do to get help move this along. I would really like to have this integration.

G-Two commented:
@Geezus42 thanks for the interest!  I'm excited to have the integration added.  It looks like this PR is getting close to the front of the queue.  While we wait, I welcome any additional review of either this integration, or the supporting [subarulink](https://github.com/G-Two/subarulink) PyPI package.  There's a standalone command line application included with `subarulink`.  If you could test it out on your vehicle and provide any feedback, that would be very helpful.  Subaru's API is not very consistent, so additional feedback from the owners of different Subaru models is very helpful.  Thanks again.

jonslilbro commented:
@G-Two I (among many others I'm sure) would love to help test. Can you give us a small write up on what the process is to get this installed/setup?

G-Two commented:
@jonslilbro great!  Here are some notes to get started with using the the module on its own:
- The easiest way to install is through PyPI with `pip install subarulink`.  The current version is subarulink==0.3.2 which has some feature additions to the standalone CLI application.  Alternatively, you can also clone the [repo](https://github.com/G-Two/subarulink), and use `pipenv` to install the dependencies.
- Once installed, you should be able to run `subarulink`.  When run without any arguments, you'll receive the help/usage instructions.
- To associate your MySubaru account, you'll need to run interactive mode `subarulink -i -v2` (interactive with debug verbosity) and enter your username/password/pin.  This will create a config file called  `~/.subarulink.cfg`.  You can change this with the `-c` option if you wish.  The file is a simple text JSON file.  Be advised that it will contain your MySubaru account credentials (similar to how HomeAssistant will keep them in `.storage`).  You also have the option to not save the account credentials, but you will then need to enter them each time.  I was thinking of a middle option where the pin number needs to be entered each time.
- After the application authenticates to MySubaru (it will take around 10-12 seconds for the first time when it adds itself as an authorized device), you'll be presented with a vehicle selection prompt if you have multiple vehicles.  If you only have one vehicle, you'll see a prompt with your vehicle name.  The commands a fairly self-explanatory.  Newer "Generation 2" telematics vehicles will have the ability to view some status information.
- The CLI application is fairly crude, as my main goal was HomeAssistant integration.  However, the CLI is useful for exercising a lot of the backend code that HomeAssistant will use.  So if you find a bug or issue affecting the CLI, chances are it is something that will need to be fixed for HomeAssistant.
- Remote actuation commands such as lock, lights, horn generally take about 10 seconds to flow to your vehicle.
- If you allow your credentials to be saved to the config file, you can also use the single command mode, which lets you issue a single command (e.g. `subarulink lock`), which could be useful for scripting. 
- Please see the [README](https://github.com/G-Two/subarulink/blob/master/README.md) for known issues and [DEVELOPERS](https://github.com/G-Two/subarulink/blob/master/DEVELOPERS.md) for information on how to use the module in your own applications.
- If you encounter any issues, please open an issue in the subarulink repo and I'll try to address them.  If you can include any log output (please use the `-v2` verbosity option), that would be helpful.  Be sure to sanitize any log data you include of personal information (the Subaru API returns things like your home address).
- Thanks!

jonslilbro commented:
@G-Two so far, working good. Have to play more with it tomorrow, but love the fact that it's connected at least. Questions for ya: is remote start a feature that's capable of being added at some point? Also, can fuel level be pulled instead of miles left? Also, any way to pull if the car is actively running or not? Great work!

G-Two committed:
Bump to subarulink==0.3.3

G-Two commented:
> @G-Two so far, working good. Have to play more with it tomorrow, but love the fact that it's connected at least. Questions for ya: is remote start a feature that's capable of being added at some point? Also, can fuel level be pulled instead of miles left? Also, any way to pull if the car is actively running or not? Great work!

@jonslilbro yes, remote start is supported by the underlying subarulink package.  I just pushed updates (v0.3.3) to PyPI that fixed some issues with its reliability.  I plan to add remote start to HA as a follow on PR after this initial one is completed.  The Subaru API returns distance to empty rather than a fuel level.  You could calculate an estimated fuel level from that and the average fuel consumption.  It does look like it is possible to get the ignition state of the vehicle from 'VEHICLE_STATE_TYPE'.

G-Two committed:
Bump to subarulink 0.3.4

- Adds support for Safety Plus subscriptions
- Adds tire pressure sensors
- Simplifies sensor.py

OnFreund commented on homeassistant/components/subaru/__init__.py:
```diff
@@ -0,0 +1,250 @@
+
+async def async_setup_entry(hass, entry):
+    """Set up Subaru from a config entry."""
+    hass.data.setdefault(DOMAIN, {})
```
```suggestion
```

OnFreund commented on homeassistant/components/subaru/__init__.py:
```diff
@@ -0,0 +1,250 @@
+        )
+        if not await controller.connect():
+            _LOGGER.error("Failed to connect")
+            return False
```
```suggestion
            raise ConfigEntryNotReady
```

OnFreund commented on homeassistant/components/subaru/__init__.py:
```diff
@@ -0,0 +1,250 @@
+            return False
+    except SubaruException as ex:
+        _LOGGER.error("Unable to communicate with Subaru API: %s", ex.message)
+        return False
```
```suggestion
        raise ConfigEntryNotReady from ex
```

OnFreund commented on homeassistant/components/subaru/__init__.py:
```diff
@@ -0,0 +1,250 @@
+                data = await subaru_update(vehicle_info, controller)
+            return data
+        except SubaruException as err:
+            raise UpdateFailed(f"Error communicating with API: {err}")
```
```suggestion
            raise ConfigEntryNotReady from err
```

OnFreund commented on homeassistant/components/subaru/__init__.py:
```diff
@@ -0,0 +1,250 @@
+async def update_listener(hass, config_entry):
+    """Update when config_entry options update."""
+    controller = hass.data[DOMAIN][config_entry.entry_id]["controller"]
+    controller.set_update_interval(config_entry.options.get(CONF_HARD_POLL_INTERVAL))
+    controller.set_fetch_interval(config_entry.options.get(CONF_SCAN_INTERVAL))
```
You should use defaults in case these don't exist

G-Two replied:
defaults added in 934d31e

OnFreund commented on homeassistant/components/subaru/__init__.py:
```diff
@@ -0,0 +1,250 @@
+    controller.set_fetch_interval(config_entry.options.get(CONF_SCAN_INTERVAL))
+
+
+class SubaruEntity(Entity):
```
Probably belongs in its own file

G-Two replied:
change made in 934d31e

OnFreund commented on homeassistant/components/subaru/__init__.py:
```diff
@@ -0,0 +1,250 @@
+    @property
+    def unique_id(self) -> str:
+        """Return a unique ID."""
+        return slugify(
```
Shouldn't the vin be enough?

G-Two replied:
Yes, I agree.  Change made in 934d31e.

OnFreund commented on homeassistant/components/subaru/__init__.py:
```diff
@@ -0,0 +1,250 @@
+    @property
+    def should_poll(self):
+        """Return the polling state."""
+        return self._should_poll
```
```suggestion
        return False
```

OnFreund commented on homeassistant/components/subaru/__init__.py:
```diff
@@ -0,0 +1,250 @@
+        self.car_name = vehicle_info["display_name"]
+        self.vin = vehicle_info["vin"]
+        self._attributes = {}
+        self._should_poll = False
```
```suggestion
```

OnFreund commented on homeassistant/components/subaru/__init__.py:
```diff
@@ -0,0 +1,250 @@
+            self.coordinator.async_add_listener(self.async_write_ha_state)
+        )
+
+    async def async_will_remove_from_hass(self):
```
You can remove this if you're not doing anything in it

G-Two replied:
Removed in 934d31e

OnFreund commented on homeassistant/components/subaru/__init__.py:
```diff
@@ -0,0 +1,250 @@
+        self.coordinator = coordinator
+        self.car_name = vehicle_info["display_name"]
+        self.vin = vehicle_info["vin"]
+        self._attributes = {}
```
```suggestion
```

OnFreund commented on homeassistant/components/subaru/__init__.py:
```diff
@@ -0,0 +1,250 @@
+        """Return the polling state."""
+        return self._should_poll
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes of the device."""
+        return self._attributes
```
```suggestion
```

OnFreund commented on homeassistant/components/subaru/__init__.py:
```diff
@@ -0,0 +1,250 @@
+
+    for vin in vehicle_info.keys():
+        # Only g2 api vehicles have data updates
+        if vehicle_info[vin]["api_gen"] == "g2":
```
Maybe just move to next vehicle if not g2? Will be easier to follow the logic.
I would also refactor this into a function

G-Two replied:
Agreed.  Changes made in 934d31e

OnFreund commented on homeassistant/components/subaru/__init__.py:
```diff
@@ -0,0 +1,250 @@
+
+def get_vehicle_info(controller, vin):
+    """Obtain vehicle identifiers and capabilities."""
+    info = {}
+    info["vin"] = vin
+    info["display_name"] = controller.vin_to_name(vin)
+    info["is_ev"] = controller.get_ev_status(vin)
+    info["api_gen"] = controller.get_api_gen(vin)
+    info["has_res"] = controller.get_res_status(vin)
+    info["has_remote"] = controller.get_remote_status(vin)
+    info["has_safety"] = controller.get_safety_status(vin)
+    info["last_update"] = 0
+    info["last_fetch"] = 0
```
```suggestion
    info = {
        "vin": vin,
        "display_name": controller.vin_to_name(vin),
        "is_ev": controller.get_ev_status(vin),
        "api_gen": controller.get_api_gen(vin),
        "has_res": controller.get_res_status(vin),
        "has_remote": controller.get_remote_status(vin),
        "has_safety": controller.get_safety_status(vin),
        "last_update": 0,
        "last_fetch": 0,
    }
```

OnFreund commented on homeassistant/components/subaru/config_flow.py:
```diff
@@ -0,0 +1,170 @@
+            )
+
+        if user_input[CONF_USERNAME] in configured_instances(self.hass):
+            return self.async_show_form(
```
Should this case abort?

G-Two replied:
Yes, change made in 934d31e.

OnFreund commented on homeassistant/components/subaru/config_flow.py:
```diff
@@ -0,0 +1,170 @@
+        try:
+            info = await validate_input(self.hass, user_input)
+        except CannotConnect:
+            return self.async_show_form(
+                step_id="user",
+                data_schema=DATA_SCHEMA,
+                errors={"base": "connection_error"},
+                description_placeholders={},
+            )
+        except InvalidAuth:
+            return self.async_show_form(
+                step_id="user",
+                data_schema=DATA_SCHEMA,
+                errors={"base": "invalid_credentials"},
+                description_placeholders={},
```
When showing the form again, it'll probably be a better user experience to pre-populate the fields with the previous values.

G-Two replied:
Form is now pre-populated for both invalid password and invalid PIN errors.

OnFreund commented on homeassistant/components/subaru/config_flow.py:
```diff
@@ -0,0 +1,170 @@
+        if await controller.connect(test_login=True):
+            _LOGGER.debug("Successfully authenticated and authorized with Subaru API")
+        else:
+            raise SubaruException("Unknown Error")
```
Since this will be caught a couple of lines down and raised as a `CannotConnect`, why not simply raise `CannotConnect` here?

G-Two replied:
Agreed, config flow exceptions were redone to just make use of what is provided by the subarulink module.

OnFreund commented on homeassistant/components/subaru/config_flow.py:
```diff
@@ -0,0 +1,170 @@
+    except SubaruException as ex:
+        if ex.message == "invalidAccount":
+            _LOGGER.error("Invalid credentials: %s", ex)
+            raise InvalidAuth()
```
```suggestion
            raise InvalidAuth
```

OnFreund commented on homeassistant/components/subaru/config_flow.py:
```diff
@@ -0,0 +1,170 @@
+            _LOGGER.error("Invalid credentials: %s", ex)
+            raise InvalidAuth()
+        _LOGGER.error("Unable to communicate with Subaru API: %s", ex)
+        raise CannotConnect()
```
```suggestion
        raise CannotConnect
```

OnFreund commented on homeassistant/components/subaru/const.py:
```diff
@@ -0,0 +1,29 @@
+MIN_HARD_POLL_INTERVAL = 300
+CONF_HARD_POLL_INTERVAL = "hard_poll_interval"
+
+SUPPORTED_PLATFORMS = [
+    "lock",
+    "sensor",
```
New integrations should only have one platform in the initial PR. Probably best to start with sensor.

G-Two replied:
Removed lock platform for this initial PR.

OnFreund commented on homeassistant/components/subaru/lock.py:
```diff
@@ -0,0 +1,53 @@
+"""Support for Subaru door locks."""
```
Since you should only have one platform, this file should be reviewed in the future PR in which it's added.

G-Two replied:
Removed lock platform for this initial PR.

OnFreund commented on homeassistant/components/subaru/sensor.py:
```diff
@@ -0,0 +1,251 @@
+)
+from homeassistant.util.volume import convert as vol_convert
+
+from . import DOMAIN as SUBARU_DOMAIN, SubaruEntity
```
```suggestion
from . import SubaruEntity
from .const import DOMAIN as SUBARU_DOMAIN
```

OnFreund commented on homeassistant/components/subaru/sensor.py:
```diff
@@ -0,0 +1,251 @@
+
+def _create_sensor_entities(entities, vehicle_info, coordinator, hass):
+    if vehicle_info["has_safety"]:
+        entities.append(
```
Maybe have a list of sensors and iterate over it?

G-Two replied:
Refactored in 934d31e to use a list.

OnFreund commented on homeassistant/components/subaru/config_flow.py:
```diff
@@ -0,0 +1,170 @@
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_CLOUD_POLL
+
+    async def async_step_import(self, import_config):
+        """Import a config entry from configuration.yaml."""
+        return await self.async_step_user(import_config)
```
I don't think this is used anywhere

G-Two replied:
Removed in 934d31e

OnFreund commented on homeassistant/components/subaru/strings.json:
```diff
@@ -0,0 +1,31 @@
+    "step": {
+      "user": {
+        "data": {
+          "username": "Email Address",
+          "password": "Password",
```
```suggestion
          "username": "[%key:common::config_flow::data::username%]",
          "password": "[%key:common::config_flow::data::password%]",
```

OnFreund commented on homeassistant/components/subaru/strings.json:
```diff
@@ -0,0 +1,31 @@
+{
+  "config": {
+    "error": {
+      "connection_error": "Error connecting; check network and retry",
+      "identifier_exists": "Email already registered",
+      "invalid_credentials": "Invalid credentials",
+      "unknown_error": "Unknown error, please report log info"
```
```suggestion
      "cannot_connect": "[%key:common::config_flow::error::cannot_connect%]",
      "invalid_auth": "[%key:common::config_flow::error::invalid_auth%]",
      "unknown": "[%key:common::config_flow::error::unknown%]"
      "already_configured": "[%key:common::config_flow::abort::already_configured_device%]"
```

And, as mentioned previously, the `already_configured` case might be an abort.

OnFreund replied:
Also, where is `unknown_error` used?

OnFreund commented on homeassistant/components/subaru/config_flow.py:
```diff
@@ -0,0 +1,170 @@
+            return self.async_show_form(
+                step_id="user",
+                data_schema=DATA_SCHEMA,
+                errors={"base": "invalid_credentials"},
```
```suggestion
                errors={"base": "invalid_auth"},
```

OnFreund commented on homeassistant/components/subaru/config_flow.py:
```diff
@@ -0,0 +1,170 @@
+            return self.async_show_form(
+                step_id="user",
+                data_schema=DATA_SCHEMA,
+                errors={"base": "connection_error"},
```
```suggestion
                errors={"base": "cannot_connect"},
```

OnFreund commented on homeassistant/components/subaru/config_flow.py:
```diff
@@ -0,0 +1,170 @@
+            return self.async_show_form(
+                step_id="user",
+                data_schema=DATA_SCHEMA,
+                errors={"base": "identifier_exists"},
```
```suggestion
                errors={"base": "already_configured"},
```

OnFreund commented on homeassistant/components/subaru/translations/en.json:
```diff
@@ -0,0 +1,32 @@
+{
```
This file shouldn't be committed. It will be created automatically.

G-Two replied:
File removed.

OnFreund commented on tests/components/subaru/test_config_flow.py:
```diff
@@ -0,0 +1,198 @@
+        )
+
+    assert result2["type"] == "form"
+    assert result2["errors"] == {"base": "invalid_credentials"}
```
```suggestion
    assert result2["errors"] == {"base": "invalid_auth"}
```

OnFreund commented on tests/components/subaru/test_config_flow.py:
```diff
@@ -0,0 +1,198 @@
+        )
+
+    assert result2["type"] == "form"
+    assert result2["errors"] == {"base": "connection_error"}
```
```suggestion
    assert result2["errors"] == {"base": "cannot_connect"}
```

OnFreund commented on tests/components/subaru/test_config_flow.py:
```diff
@@ -0,0 +1,198 @@
+        )
+
+    assert result2["type"] == "form"
+    assert result2["errors"] == {"base": "identifier_exists"}
```
```suggestion
    assert result2["errors"] == {"base": "already_configured"}
```

ianwestcott commented:
@G-Two I'm eager to see this merged. Is there anything I can do to help you push it over the line?

G-Two commented:
> @G-Two I'm eager to see this merged. Is there anything I can do to help you push it over the line?

@ianwestcott thanks for the interest.  I wasn't able to devote any time to this over the past couple months.  I finally have some free time now, and I think I can incorporate all of the requested changes over this weekend.

G-Two committed:
Apply suggestions from code review

Co-authored-by: On Freund <onfreund@gmail.com>

G-Two committed:
Applies all suggestions from code review

- Also adds PIN validation during config flow
- Improves exception handling
- Bumps to subarulink 0.3.6

G-Two commented on homeassistant/components/subaru/__init__.py:
```diff
@@ -0,0 +1,250 @@
+    async def async_update_data():
+        """Fetch data from API endpoint."""
+        try:
+            async with async_timeout.timeout(30):
```
Removed two extraneous timeouts/intervals in 934d31e
- Timeout for fetching data removed (now responsibility of subarulink module to throw exception)
- Separate coordinator update interval removed (data coordinator update is set to scan interval now)

G-Two commented on homeassistant/components/subaru/__init__.py:
```diff
@@ -0,0 +1,250 @@
+    await coordinator.async_refresh()
+
+    hass.data[DOMAIN][entry.entry_id] = {
+        "controller": controller,
```
dict keys now use consts in 934d31e

G-Two commented on homeassistant/components/subaru/__init__.py:
```diff
@@ -0,0 +1,250 @@
+        "controller": controller,
+        "coordinator": coordinator,
+        "vehicles": vehicle_info,
+        "listener": [entry.add_update_listener(update_listener)],
```
change made in 934d31e

G-Two committed:
Update sensor.py to fix pylint error

OnFreund commented on homeassistant/components/subaru/__init__.py:
```diff
@@ -0,0 +1,232 @@
+        try:
+            return await subaru_update(vehicle_info, controller)
+        except InvalidPIN as err:
+            _LOGGER.error(
```
You shouldn't have your own error logs, since the updater takes care of only printing once instead of filling the logs with them.
I would also change to `raise UpdateFailed(err) from err`, otherwise the updater log won't contain any specific information.

G-Two replied:
Changes made.

OnFreund commented on homeassistant/components/subaru/__init__.py:
```diff
@@ -0,0 +1,232 @@
+            )
+            raise UpdateFailed from err
+        except PINLockoutProtect as err:
+            _LOGGER.error(
```
same here

G-Two replied:
Changes made

OnFreund commented on homeassistant/components/subaru/config_flow.py:
```diff
@@ -0,0 +1,168 @@
+                errors=error,
+            )
+
+        return self.async_create_entry(title=user_input[CONF_USERNAME], data=info)
```
You can do this in an `else:` after the `try:`

G-Two replied:
Change made.

OnFreund commented on homeassistant/components/subaru/config_flow.py:
```diff
@@ -0,0 +1,168 @@
+            {
+                vol.Optional(
+                    CONF_SCAN_INTERVAL,
+                    default=self.config_entry.options.get(
```
If you're setting a default this isn't really optional :)

G-Two replied:
Change made.

OnFreund commented on homeassistant/components/subaru/config_flow.py:
```diff
@@ -0,0 +1,168 @@
+
+    _LOGGER.debug("Testing user provided PIN with Subaru remote service requests")
+    if await controller.test_pin():
+        _LOGGER.debug("User provided PIN is valid for Subaru remote service requests")
```
This branch isn't covered by your tests

G-Two replied:
Added a bunch more tests.

OnFreund commented on homeassistant/components/subaru/entity.py:
```diff
@@ -0,0 +1,60 @@
+        return ICONS.get(self.title)
+
+    @property
+    def should_poll(self):
```
You can inherit from `CoordinatorEntity` which would take care of the boilerplate for you.

G-Two replied:
Thanks! Change made.

OnFreund commented on homeassistant/components/subaru/entity.py:
```diff
@@ -0,0 +1,60 @@
+        """Return if entity is available."""
+        return self.coordinator.last_update_success
+
+    async def async_added_to_hass(self):
+        """Register state update callback."""
+        self.async_on_remove(
+            self.coordinator.async_add_listener(self.async_write_ha_state)
+        )
+
+    async def async_update(self):
+        """Update the state of the device."""
+        await self.coordinator.async_request_refresh()
```
This will also be made redundant by inhering from `CoordinatorEntity`

G-Two replied:
Change made.

OnFreund commented on homeassistant/components/subaru/manifest.json:
```diff
@@ -0,0 +1,12 @@
+  "config_flow": true,
+  "documentation": "https://www.home-assistant.io/integrations/subaru",
+  "requirements": ["subarulink==0.3.6"],
+  "ssdp": [],
+  "zeroconf": [],
+  "homekit": {},
+  "dependencies": [],
```
```suggestion
```

G-Two replied:
Change made.

G-Two committed:
Incorporates additional code review changes

- Also adds more tests

OnFreund commented on homeassistant/components/subaru/__init__.py:
```diff
@@ -0,0 +1,250 @@
+
+async def async_setup(hass, base_config):
+    """Set up of Subaru component."""
+    return True
```
looks like you accidentally removed this.

G-Two replied:
Fixed

OnFreund commented on homeassistant/components/subaru/__init__.py:
```diff
@@ -74,11 +68,12 @@ async def async_setup_entry(hass, entry):
+    except SubaruException as err:
+        raise ConfigEntryNotReady(err) from err
+
+    if DOMAIN not in hass.data:
```
not needed after you add this again to `async_setup`

G-Two replied:
Fixed

OnFreund commented on homeassistant/components/subaru/entity.py:
```diff
@@ -48,13 +43,3 @@ def device_info(self):
     def available(self):
```
I think you can also remove this when inheriting from `CoordinatorEntity`

G-Two replied:
Removed.

OnFreund commented on homeassistant/components/subaru/config_flow.py:
```diff
@@ -0,0 +1,167 @@
+        """Handle the start of the config flow."""
+        error = None
+
+        if not user_input:
+            return self.async_show_form(
+                step_id="user",
+                data_schema=vol.Schema(
+                    {
+                        vol.Required(CONF_USERNAME): str,
+                        vol.Required(CONF_PASSWORD): str,
+                        vol.Required(CONF_PIN): str,
+                    }
+                ),
+            )
```
This isn't needed anymore - you reverse the condition, and then use a single show form (with or without errors) at the bottom

G-Two replied:
Change incorporated

G-Two committed:
Incorporates additional code review changes

G-Two committed:
Adds test coverage to sensor.py

OnFreund reviewed (approved):
LGTM. Keep in mind this still needs to be reviewed by the core team.

G-Two commented:
@OnFreund thanks for the review!

dorkmatt commented:
@G-Two Minor note, the `requirements` file is against the 0.3.8 library version - however it looks like you've only released up to 0.3.7 :smiley:  Sadly even changing this to released version and placing the code under `custom_components` didn't seem to work. Will wait until this is formally released, thank you for writing this!

G-Two committed:
Bump subarulink to 0.3.8

G-Two commented:
@dorkmatt Thanks for the note.  I've released 0.3.8 and deleted the extraneous branches in my fork. To help keep things separate while I wait for this review, I've moved all new feature development beyond this initial PR to a [separate repo](https://github.com/G-Two/homeassistant-subaru).

dorkmatt commented:
@G-Two In my prior testing, I attempted to use the older YAML style configuration approach. The UI seems to work, but none of the fields have names (see attachment).
![image](https://user-images.githubusercontent.com/139549/100530967-6fd0f300-31ad-11eb-8947-ba3188fc19f5.png)
Also, guessing something might have changed on their API backend? As I get an "Aborted" UI message when attempting to authenticate.

```
2020-11-28 19:03:18 INFO (MainThread) [custom_components.subaru.config_flow] Setting up first time connection to Subuaru API.  This may take up to 20 seconds.
2020-11-28 19:03:23 INFO (MainThread) [subarulink.connection] Device successfully authorized
2020-11-28 19:03:33 DEBUG (MainThread) [custom_components.subaru.config_flow] Successfully authenticated and authorized with Subaru API
2020-11-28 19:03:33 DEBUG (MainThread) [custom_components.subaru.config_flow] Testing user provided PIN with Subaru remote service requests
2020-11-28 19:03:33 INFO (MainThread) [subarulink.controller] Testing PIN for validity with Subaru remote services
2020-11-28 19:03:33 ERROR (MainThread) [subarulink.connection] RequestInfo(url=URL('https://mobileapi.prod.subarucs.com/g2v17/service/g2/vehicleStatus/execute.json'), method='POST', headers=<CIMultiDictProxy('Host': 'mobileapi.prod.subarucs.com', 'User-Agent': 'Mozilla/5.0 (Linux; Android 10; Android SDK built for x86 Build/QSR1.191030.002; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/74.0.3729.185 Mobile Safari/537.36', 'Origin': 'file://', 'X-Requested-With': 'com.subaru.telematics.app.remote', 'Accept-Language': 'en-US,en;q=0.9', 'Accept-Encoding': 'gzip, deflate', 'Accept': '*/*', 'Cookie': 'JSESSIONID=XXX; ORA_OTD_JROUTE=XXX', 'Content-Length': '15', 'Content-Type': 'application/json')>, real_url=URL('https://mobileapi.prod.subarucs.com/g2v17/service/g2/vehicleStatus/execute.json'))
2020-11-28 19:03:33 ERROR (MainThread) [subarulink.connection] ('{"httpCode":500,"errorCode":"error","errorMessage":"java.lang.NullPointerException '
2020-11-28 19:03:33 ERROR (MainThread) [custom_components.subaru.config_flow] Unable to communicate with Subaru API: HTTP 500: <ClientResponse(https://mobileapi.prod.subarucs.com/g2v17/service/g2/vehicleStatus/execute.json) [500 Server Error]>
```

G-Two commented:
@dorkmatt hmm...there was [another user](https://github.com/G-Two/subarulink/issues/23) that experienced the exact same HTTP 500 issue.  It's clearly something that happens in `subarulink.controller.test_pin()`, but I don't understand why.  Can you check your log for the string "Parsing vehicle:" and let me know how many different VINs show up?  Sometimes an old VIN stays on your account.  I've tried adding checks for those and ignoring them, but I might be making a bad assumption on them.  Could you also add `subarulink: debug` to your HA logging options? 

As for the empty fields in the config flow UI, I think that is an artifact of the automated translations for Home Assistant Core.  We can't commit translation json files anymore, as they are "compiled" in an automated fashion later.  You can always pull from the [standalone repo](https://github.com/G-Two/homeassistant-subaru) I setup, as that has the english translation files included (along with more features that need user testing).

dorkmatt commented:
@G-Two Sure, additional logs below _(sanitized)_. Note the standalone CLI tool included in the library works fine.

```
2020-11-28 23:30:43 INFO (MainThread) [custom_components.subaru.config_flow] Setting up first time connection to Subuaru API.  This may take up to 20 seconds.
2020-11-28 23:30:43 DEBUG (MainThread) [subarulink.controller] Connecting controller to Subaru Remote Services
2020-11-28 23:30:43 DEBUG (MainThread) [subarulink.connection] POST: https://mobileapi.prod.subarucs.com/g2v17/login.json
2020-11-28 23:30:44 DEBUG (MainThread) [subarulink.connection] Client authentication successful
2020-11-28 23:30:44 DEBUG (MainThread) [subarulink.connection] {'data': {'account': {'accountKey': XXX,
2020-11-28 23:30:44 DEBUG (MainThread) [subarulink.connection] GET: https://mobileapi.prod.subarucs.com/g2v17/refreshVehicles.json
2020-11-28 23:30:45 DEBUG (MainThread) [subarulink.connection] {'data': {'account': {'accountKey': XXX,
2020-11-28 23:30:45 DEBUG (MainThread) [subarulink.connection] Authorizing device via web API
2020-11-28 23:30:45 DEBUG (MainThread) [subarulink.connection] POST: https://www.mysubaru.com/login
2020-11-28 23:30:48 DEBUG (MainThread) [subarulink.connection] GET: https://www.mysubaru.com/profile/updateDeviceEntry.json
2020-11-28 23:30:48 INFO (MainThread) [subarulink.connection] Device successfully authorized
2020-11-28 23:30:48 DEBUG (MainThread) [subarulink.connection] Setting Device Name to Home Assistant: Added 2020-11-28
2020-11-28 23:30:48 DEBUG (MainThread) [subarulink.connection] GET: https://www.mysubaru.com/profile/addDeviceName.json
2020-11-28 23:30:48 DEBUG (MainThread) [subarulink.connection] Set Device Name Successful
2020-11-28 23:30:51 DEBUG (MainThread) [subarulink.connection] POST: https://mobileapi.prod.subarucs.com/g2v17/login.json
2020-11-28 23:30:52 DEBUG (MainThread) [subarulink.connection] Client authentication successful
2020-11-28 23:30:52 DEBUG (MainThread) [subarulink.connection] {'data': {'account': {'accountKey': XXX,
2020-11-28 23:30:55 DEBUG (MainThread) [subarulink.connection] POST: https://mobileapi.prod.subarucs.com/g2v17/login.json
2020-11-28 23:30:55 DEBUG (MainThread) [subarulink.connection] Client authentication successful
2020-11-28 23:30:55 DEBUG (MainThread) [subarulink.connection] {'data': {'account': {'accountKey': XXX,
2020-11-28 23:30:55 DEBUG (MainThread) [subarulink.controller] Parsing vehicle: 4S4xxx
2020-11-28 23:30:55 DEBUG (MainThread) [subarulink.controller] Subaru Remote Services Ready!
2020-11-28 23:30:55 DEBUG (MainThread) [custom_components.subaru.config_flow] Successfully authenticated and authorized with Subaru API
2020-11-28 23:30:55 DEBUG (MainThread) [custom_components.subaru.config_flow] Testing user provided PIN with Subaru remote service requests
2020-11-28 23:30:55 INFO (MainThread) [subarulink.controller] Testing PIN for validity with Subaru remote services
2020-11-28 23:30:55 DEBUG (MainThread) [subarulink.controller] Getting subscription Status 4S4xxx:True
2020-11-28 23:30:55 DEBUG (MainThread) [subarulink.controller] Getting remote Status 4SBxxx:True
2020-11-28 23:30:55 DEBUG (MainThread) [subarulink.connection] GET: https://mobileapi.prod.subarucs.com/g2v17/validateSession.json
2020-11-28 23:30:55 DEBUG (MainThread) [subarulink.connection] {'data': None, 'dataName': None, 'errorCode': None, 'success': True}
2020-11-28 23:30:55 DEBUG (MainThread) [subarulink.controller] Getting vehicle API gen 4S4xxx:g2
2020-11-28 23:30:55 DEBUG (MainThread) [subarulink.connection] POST: https://mobileapi.prod.subarucs.com/g2v17/service/g2/vehicleStatus/execute.json
2020-11-28 23:30:55 ERROR (MainThread) [subarulink.connection] RequestInfo(url=URL('https://mobileapi.prod.subarucs.com/g2v17/service/g2/vehicleStatus/execute.json'), method='POST', headers=<CIMultiDictProxy('Host': 'mobileapi.prod.subarucs.com', 'User-Agent': 'Mozilla/5.0 (Linux; Android 10; Android SDK built for x86 Build/QSR1.191030.002; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/74.0.3729.185 Mobile Safari/537.36', 'Origin': 'file://', 'X-Requested-With': 'com.subaru.telematics.app.remote', 'Accept-Language': 'en-US,en;q=0.9', 'Accept-Encoding': 'gzip, deflate', 'Accept': '*/*', 'Cookie': 'JSESSIONID=1E4A4013FF4EE893BB0761DEAF64FF86; ORA_OTD_JROUTE=0t-ZDpuqZ8Kd9fOPUgWG3z-Lj7lLqZqf', 'Content-Length': '15', 'Content-Type': 'application/json')>, real_url=URL('https://mobileapi.prod.subarucs.com/g2v17/service/g2/vehicleStatus/execute.json'))
2020-11-28 23:30:55 ERROR (MainThread) [subarulink.connection] ('{"httpCode":500,"errorCode":"error","errorMessage":"java.lang.NullPointerException '
2020-11-28 23:30:55 ERROR (MainThread) [custom_components.subaru.config_flow] Unable to communicate with Subaru API: HTTP 500: <ClientResponse(https://mobileapi.prod.subarucs.com/g2v17/service/g2/vehicleStatus/execute.json) [500 Server Error]>
```

G-Two commented:
@dorkmatt thanks for the logs.  My theory is that after authentication, the current VIN context on Subaru's backend is undefined when we try to run the `test_pin()` function, and the controller is assuming that its current VIN is in sync with the backend without explicitly setting it with `_select_vehicle()`.  I didn't think this was possible, especially if only one VIN is on an account.  I will make a change to no longer use the API reported value for `currentVehicleIndex` as a basis for the current VIN context and instead set the initial value to `None`.  This will force a call to `_select_vehicle(vin)` during `test_pin()` and will hopefully fix the problem.

G-Two committed:
Bump to subarulink 0.3.9

dorkmatt commented:
@G-Two The 0.3.9 version appears to work!
![image](https://user-images.githubusercontent.com/139549/100555489-a875d880-3250-11eb-8b94-fe46f65fba0e.png)
Are you planning to add device tracker support?

stboch commented:
@dorkmatt pull from the other standalone repo, inital home-assistant new integration only like you implementing 1 domain at the first PR. Additionally the location only updates if you force a location update or when the cars engine turns off.

jeffphilly commented:
Recently tried adding this with my 2017 Forrester. Not sure what is goin wrong but I am not getting any entities or devices showing up on the integration. Was able to get through the email, pw and pin, but then nothing shows other then the blank integration. Hopefully I can help test and troubleshoot, let me know what i can get you to help figure it out.

G-Two commented:
@jeffphilly the problem is likely with the subarulink module.  It hasn't been tested with a "generation 1" telematics vehicle (model year 2016-2018) vehicle yet.  The API is different, so I made some best effort guesses, but need more sample response data to fully support.  Please open an issue [here](https://github.com/G-Two/subarulink).  If you can, follow the instructions for the standalone CLI application and paste the debug logs after removing personal info.

tango2590 commented:
I apologize in advance, I'm pretty new to GitHub and python. But I'm also pretty invested in this project and want to help as much as I can. Currently trying to work out the basics in Terminal following an older comment's explanation, but I'm running into an issue. Config file was successfully created but it keeps failing to connect.

>ryan@ryan ~ % subarulink -i -v2
2021-02-06 17:52:35,336 - subarulink - INFO - Entering interactive mode
2021-02-06 17:52:35,336 - subarulink - INFO - Opening config file: /Users/ryan/.subarulink.cfg
2021-02-06 17:52:35,337 - subarulink - INFO - Connecting to Subaru Remote Services API
2021-02-06 17:52:35,337 - subarulink.controller - DEBUG - subarulink 0.3.12
2021-02-06 17:52:35,337 - subarulink.controller - DEBUG - Connecting controller to Subaru Remote Services
2021-02-06 17:52:35,337 - subarulink.connection - DEBUG - POST: https://mobileapi.prod.subarucs.com/g2v17/login.json, params=None, json_data=None
2021-02-06 17:52:35,383 - subarulink - ERROR - Unable to connect: aiohttp.ClientConnectionError

The json link says "invalid account" for the error code. I double checked the credentials in the config file and they're correct. It's a 2020 vehicle with the full safety package. Would love to utilize this in my automation project. Any help is appreciated.

G-Two commented:
@tango2590 please [start a discussion](https://github.com/G-Two/subarulink/discussions) or [open a new issue](https://github.com/G-Two/subarulink/issues) in the main subarulink module repo. Based on your log output, it doesn't look like a problem related to the Home Assistant integration or this PR in particular.

bdraco commented on homeassistant/components/subaru/entity.py:
```diff
@@ -0,0 +1,40 @@
+    @property
+    def unique_id(self) -> str:
+        """Return a unique ID."""
+        return slugify(f"{DOMAIN} {self.vin} {self.title}")
```
We don't need the domain in the unique id

Unique ids need to be unique to the platform, domain, unique id itself.

G-Two replied:
domain removed

bdraco commented on homeassistant/components/subaru/entity.py:
```diff
@@ -0,0 +1,40 @@
+    @property
+    def unique_id(self) -> str:
+        """Return a unique ID."""
+        return slugify(f"{DOMAIN} {self.vin} {self.title}")
```
```suggestion
        return f"{self.vin}_{self.title}"
```

Is title stable?

G-Two replied:
Yes, title is stable.

bdraco replied:
Please rename `self.title` to something that implies it is not configurable by the user.  Right now `title` implies its a name, which will make this more difficult to review in the future as it will trigger the reviewer to check to make sure its a stable identifier.

Maybe `self.XXXX_type`

G-Two replied:
changed to `self.entity_type`

bdraco commented on homeassistant/components/subaru/__init__.py:
```diff
@@ -0,0 +1,212 @@
+
+async def update_listener(hass, config_entry):
+    """Update when config_entry options update."""
+    controller = hass.data[DOMAIN][config_entry.entry_id][ENTRY_CONTROLLER]
+    coordinator = hass.data[DOMAIN][config_entry.entry_id][ENTRY_COORDINATOR]
```
```suggestion
    data = hass.data[DOMAIN][config_entry.entry_id]
    controller = data[ENTRY_CONTROLLER]
    coordinator = data[ENTRY_COORDINATOR]
```

bdraco commented on homeassistant/components/subaru/__init__.py:
```diff
@@ -0,0 +1,212 @@
+    """Set up Subaru from a config entry."""
+    config = entry.data
+    websession = aiohttp_client.async_get_clientsession(hass)
+    date = datetime.now().strftime("%Y-%m-%d")
+    device_name = "Home Assistant: Added " + date
```
This code is duplicated in config flow

G-Two replied:
removed from `__init__.py`

bdraco commented on homeassistant/components/subaru/__init__.py:
```diff
@@ -0,0 +1,212 @@
+    """Obtain vehicle identifiers and capabilities."""
+    info = {
+        VEHICLE_VIN: vin,
+        VEHICLE_NAME: controller.vin_to_name(vin),
```
Is this doing I/O?

G-Two replied:
No I/O for any of these calls.

bdraco commented on homeassistant/components/subaru/config_flow.py:
```diff
@@ -0,0 +1,162 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+@callback
+def configured_instances(hass):
+    """Return a set of configured Subaru instances."""
+    return {entry.title for entry in hass.config_entries.async_entries(DOMAIN)}
+
+
```
We should check for CONF_USERNAME in the entry not the title

bdraco replied:
`self._async_current_entries()` might make more sense

G-Two replied:
I'm now using `self._async_current_entries()`, thanks for the suggestion

bdraco commented on homeassistant/components/subaru/sensor.py:
```diff
@@ -0,0 +1,232 @@
+        super().__init__(vehicle_info, coordinator)
+        self.hass_type = "sensor"
+        self.current_value = None
+        self.hass = hass
```
Don't pass it hass, it will be set when the entity is added

G-Two replied:
removed

G-Two committed:
Apply suggestions from code review

Co-authored-by: J. Nick Koston <nick@koston.org>

G-Two committed:
Apply suggestions from code review

- Bump to subarulink 0.3.12

G-Two committed:
Fix pylint isort errors and rebase repo

G-Two committed:
Refactor tests

G-Two committed:
Enable more sensors for Safety+ subscribers

G-Two committed:
Update config flow

* Support users with Canada accounts
* Prompt for PIN only if necessary

G-Two commented:
@bdraco thanks for the review. I've incorporated all your recommendations. I hope you don't mind reviewing again with a couple commits that I added that change the config flow to support users in Canada, and some very minor changes to the sensors. 

I've also been having a hard time passing isort CI. I've rebased and am running in a fresh dev container using pre-commit. I've tried manually invoking isort in this environment, and am unable to have the subarulink imports occur before the homeassistant ones.
```
➜  home-assistant git:(subaru_initial_pr) ✗ isort --profile black homeassistant/components/subaru
➜  home-assistant git:(subaru_initial_pr) ✗ pre-commit run --hook-stage manual isort --all-files --show-diff-on-failure
isort....................................................................Passed
➜  home-assistant git:(subaru_initial_pr) ✗
 ```

G-Two committed:
Fix isort errors

bdraco commented on homeassistant/components/subaru/entity.py:
```diff
@@ -0,0 +1,39 @@
+        return {
+            "identifiers": {(DOMAIN, self.vin)},
+            "name": self.car_name,
+            "manufacturer": DOMAIN,
```
`manufacturer` is human readable so it should be a proper name

G-Two replied:
updated

bdraco commented on homeassistant/components/subaru/__init__.py:
```diff
@@ -0,0 +1,213 @@
+
+    await coordinator.async_refresh()
+
+    hass.data.get(DOMAIN)[entry.entry_id] = {
```
```suggestion
    hass.data[DOMAIN][entry.entry_id] = {
```

G-Two replied:
updated

bdraco commented on homeassistant/components/subaru/__init__.py:
```diff
@@ -0,0 +1,213 @@
+    for vehicle in vehicle_info.values():
+        vin = vehicle[VEHICLE_VIN]
+
+        # Active subscription required
+        if not vehicle[VEHICLE_HAS_SAFETY_SERVICE]:
+            continue
+
```
If their subscription lapses for a vehicle this will return data without the `vin` set which will likely result in some breakage.

Set the `available` property to `False` in the entity when this happens.

G-Two replied:
`available` property added to `SubaruSensor`

bdraco commented on homeassistant/components/subaru/sensor.py:
```diff
@@ -0,0 +1,242 @@
+            return FUEL_CONSUMPTION_L_PER_100KM
+
+        return self.api_unit
+
```
Add an available property here, check `super().available` and return False if `self.vin` not in `self.coordinator.data`

G-Two replied:
added

bdraco commented on homeassistant/components/subaru/sensor.py:
```diff
@@ -0,0 +1,242 @@
+
+    def get_current_value(self):
+        """Get raw value from the coordinator."""
+        if self.coordinator.data.get(self.vin):
```
```suggestion
        if self.vin not in self.coordinator.data:
            return None
```

Then outdent the block below

G-Two replied:
I just removed this check. With new available property added, I was unable to create a test case that would reach this line.

bdraco commented on tests/components/subaru/conftest.py:
```diff
@@ -0,0 +1,131 @@
+        vehicle_data=VEHICLE_DATA[TEST_VIN_2_EV],
+        vehicle_status=VEHICLE_STATUS_EV,
+    )
+    assert hass.data[DOMAIN][entry.entry_id]
```
Use the standard interfaces to hass.config_entries to verify its setup

G-Two replied:
updated

bdraco commented on tests/components/subaru/test_init.py:
```diff
@@ -0,0 +1,117 @@
+async def test_setup_with_no_config(hass):
+    """Test DOMAIN is empty if there is no config."""
+    assert await async_setup_component(hass, DOMAIN, {})
+    assert DOMAIN in hass.data
```
Avoid accessing `hass.data` directly in the test.

Use the standard interfaces to `hass.config_entries` to verify its setup

G-Two replied:
updated

bdraco commented on tests/components/subaru/test_init.py:
```diff
@@ -0,0 +1,117 @@
+
+async def test_setup_ev(hass, ev_entry):
+    """Test setup with an EV vehicle."""
+    assert hass.data[DOMAIN][ev_entry.entry_id]
```
Avoid accessing `hass.data` directly in the test.

Use the standard interfaces to `hass.config_entries` to verify its setup

G-Two replied:
updated

bdraco commented on tests/components/subaru/test_init.py:
```diff
@@ -0,0 +1,117 @@
+        vehicle_data=VEHICLE_DATA[TEST_VIN_3_G2],
+        vehicle_status=VEHICLE_STATUS_G2,
+    )
+    assert hass.data[DOMAIN][entry.entry_id]
```
Avoid accessing `hass.data` directly in the test.

Use the standard interfaces to `hass.config_entries` to verify its setup

G-Two replied:
updated

bdraco commented on tests/components/subaru/test_init.py:
```diff
@@ -0,0 +1,117 @@
+    entry = await setup_subaru_integration(
+        hass, vehicle_list=[TEST_VIN_1_G1], vehicle_data=VEHICLE_DATA[TEST_VIN_1_G1]
+    )
+    assert hass.data[DOMAIN][entry.entry_id]
```
Avoid accessing `hass.data` directly in the test.

Use the standard interfaces to `hass.config_entries` to verify its setup

G-Two replied:
updated

bdraco commented on tests/components/subaru/test_init.py:
```diff
@@ -0,0 +1,117 @@
+    )
+    await hass.async_block_till_done()
+
+    assert DOMAIN in hass.data
+    assert hass.data[DOMAIN] == {}
+
```
Avoid accessing `hass.data` directly in the test.

Use the standard interfaces to `hass.config_entries` to verify its setup

G-Two replied:
updated

bdraco commented on tests/components/subaru/test_init.py:
```diff
@@ -0,0 +1,117 @@
+
+async def test_update_failed(hass, ev_entry):
+    """Tests when coordinator update fails."""
+    coordinator = hass.data[DOMAIN][ev_entry.entry_id][ENTRY_COORDINATOR]
```
Don't access `coordinator` directly, call the service or use the helper to update the entity though the standard instances instead

G-Two replied:
I had a hard time using the standard interfaces/service due to the 10 second debounce interval for the `DataCoordinator`. I ended up just adding a `fetch` service to this integration to allow the user to manually refresh the data. Calling this service bypasses the debounce check and enables the test case to work. The other alternative I tried was adding `asyncio.sleep(10)` to the test case. It worked, but didn't seem ideal.

bdraco commented on tests/components/subaru/test_init.py:
```diff
@@ -0,0 +1,117 @@
+
+async def test_unload_entry(hass, ev_entry):
+    """Test that entry is unloaded."""
+    assert await subaru.async_unload_entry(hass, ev_entry)
+    assert DOMAIN in hass.data
+    assert hass.data[DOMAIN] == {}
```
Avoid accessing `hass.data` directly in the test.

Use the standard interfaces to `hass.config_entries` to verify its setup

G-Two replied:
updated

bdraco commented on tests/components/subaru/test_init.py:
```diff
@@ -0,0 +1,117 @@
+
+async def test_update_listener(hass, ev_entry):
+    """Test config options update listener."""
+    coordinator = hass.data[DOMAIN][ev_entry.entry_id][ENTRY_COORDINATOR]
+    controller = hass.data[DOMAIN][ev_entry.entry_id][ENTRY_CONTROLLER]
+    assert coordinator.update_interval == timedelta(seconds=DEFAULT_SCAN_INTERVAL)
+    assert controller.get_update_interval() == DEFAULT_HARD_POLL_INTERVAL
+
```
Same comment as above

G-Two replied:
updated

bdraco commented on tests/components/subaru/test_sensor.py:
```diff
@@ -0,0 +1,61 @@
+        return_value=VEHICLE_STATUS_EV,
+    ):
+        hass.config.units = IMPERIAL_SYSTEM
+        coordinator = hass.data[DOMAIN][ev_entry.entry_id][ENTRY_COORDINATOR]
```
Same comment as above

G-Two replied:
updated

bdraco commented on homeassistant/components/subaru/config_flow.py:
```diff
@@ -0,0 +1,179 @@
+                try:
+                    vol.Match(r"[0-9]{4}")(user_input[CONF_PIN])
+                    await self.controller.test_pin()
+                    _LOGGER.debug("PIN successfully tested")
+                    self.config_data.update(user_input)
```
Move this out of the try block into the `else` block since its not ever going to throw

G-Two replied:
updated

G-Two committed:
Apply suggestions from code review

G-Two committed:
Add services.yaml

bdraco commented on homeassistant/components/subaru/config_flow.py:
```diff
@@ -0,0 +1,179 @@
+        """Initialize options flow."""
+        self.config_entry = config_entry
+
+    async def async_step_init(self, user_input=None):
+        """Handle options flow."""
+        if user_input is not None:
+            return self.async_create_entry(title="", data=user_input)
+
+        data_schema = vol.Schema(
+            {
+                vol.Required(
+                    CONF_SCAN_INTERVAL,
+                    default=self.config_entry.options.get(
+                        CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL
+                    ),
+                ): vol.All(cv.positive_int, vol.Clamp(min=MIN_SCAN_INTERVAL)),
+                vol.Required(
+                    CONF_HARD_POLL_INTERVAL,
+                    default=self.config_entry.options.get(
+                        CONF_HARD_POLL_INTERVAL, DEFAULT_HARD_POLL_INTERVAL
+                    ),
```
As developers we should choose a reasonable default and not expose to the control unless its strictly needed as users tend to set it too low and get integrations banned from cloud services.

G-Two replied:
I'm glad you brought this up, because these options were bothering me when I typed up the documentation for them. I've set the API polling to a fixed 5 minute interval, and now expose a boolean enable/disable (default disable) option to poll the vehicle with a remote command on a fixed 120 minute interval.

G-Two committed:
Simplify integration config options

* API polling interval is fixed to 5 minutes
* Vehicle remote update is fixed to 120 minutes w/ optional enable

bdraco commented on homeassistant/components/subaru/__init__.py:
```diff
@@ -0,0 +1,189 @@
+        _LOGGER.debug("Using subarulink %s", controller.version)
+        await controller.connect()
+    except SubaruException as err:
+        raise ConfigEntryNotReady(err.message) from err
```
If the Credentials are invalid, we should log an error and return false￼ As we only want to retry when the service is not available or having issues￼

bdraco commented on homeassistant/components/subaru/sensor.py:
```diff
@@ -0,0 +1,251 @@
+    vehicle_info = hass.data[DOMAIN][config_entry.entry_id][ENTRY_VEHICLES]
+    entities = []
+    for vin in vehicle_info.keys():
+        _create_sensor_entities(entities, vehicle_info[vin], coordinator)
```
Return the entities and append them to the list instead of mutating entities in the function

bdraco commented on homeassistant/components/subaru/sensor.py:
```diff
@@ -0,0 +1,251 @@
+    @property
+    def icon(self):
+        """Return icon for sensor."""
+        if self.entity_type == "EV Battery Level" and self.coordinator.data.get(
```
Reverse the conditions and return the super then outdent the block below

bdraco replied:
It might make more sense to set a device class instead of an icon for some of these

bdraco reviewed (changes requested):
A few more cleanups and this should be good to go

G-Two committed:
Apply requested changes from review

bdraco commented on homeassistant/components/subaru/__init__.py:
```diff
@@ -0,0 +1,192 @@
+            continue
+
+        # Optionally send an "update" remote command to vehicle (throttled with update_interval)
+        if config_entry.options.get(CONF_UPDATE_ENABLED, False):
```
If the api can handle multiple requests at once make the inner block a coroutine and use asyncio.gather to call them in parallel

G-Two replied:
Unfortunately the api can only handle one request at a time.

bdraco reviewed (commented):
Looking good. Thanks!

G-Two committed:
Apply requested changes from review

bdraco commented on homeassistant/components/subaru/__init__.py:
```diff
@@ -0,0 +1,192 @@
+            hass.config_entries.async_forward_entry_setup(entry, component)
+        )
+
+    async def async_fetch_service(call):
+        """Force refresh of data from Subaru."""
+        vin = call.data[VEHICLE_VIN].upper()
+        if vin not in vehicle_info.keys():
+            hass.components.persistent_notification.async_create(
+                f"ERROR - Invalid VIN provided while calling {call.service}", "Subaru"
+            )
+        else:
+            if call.service == REMOTE_SERVICE_FETCH:
+                await coordinator.async_refresh()
+
+    hass.services.async_register(
+        DOMAIN, REMOTE_SERVICE_FETCH, async_fetch_service, schema=REMOTE_SERVICE_SCHEMA
+    )
+
```
```suggestion
```

We actually already have a service that does this `homeassistant.update_entity`: https://www.home-assistant.io/docs/scripts/service-calls/#homeassistant-services

We shouldn't reimplement it.   Since its using DataUpdateCoorinator, calling it on any entity will update them all.

G-Two committed:
Remove fetch service

bdraco reviewed (approved):
Looks good.

Can merge if CI passes.

With the size of this PR, I've likely still missed something even though I've been over it ~15 times.  Please address any additional reviewer comments in a followup PR.

G-Two commented:
@bdraco Thanks for taking the time to provide all the comments!

MartinHjelmare commented on tests/components/subaru/test_config_flow.py:
```diff
@@ -0,0 +1,250 @@
+    """Return options form for Subaru config flow."""
+    entry = MockConfigEntry(domain=DOMAIN, data={}, options=None)
+    entry.add_to_hass(hass)
+    return await hass.config_entries.options.async_init(entry.entry_id)
```
We must set up the integration before starting the options flow in tests to make sure that the config flow module is loaded.

MartinHjelmare reviewed (commented):
Please address the comments in a new PR. Thanks!

jonslilbro commented:
Is it as simple as uninstalling the official integration through home assistant and reinstalling the HACS integration if you want to get back the remote start feature? I didn't realize that it was lost in the upgrade and it's the single most used feature for me.

frenck commented:
Hi there @jonslilbro, this is a closed pull request. If you want to report a bug, please raise an issue. If you are in need for support/help, please join our Discord chat server or community forum.

Thanks 👍

G-Two commented on homeassistant/components/subaru/config_flow.py:
```diff
@@ -0,0 +1,157 @@
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_CLOUD_POLL
+    config_data = {CONF_PIN: None}
```
All comments have been incorporated in #48193. Thanks for the review!
